### DISCUSSION TITLE
Allow to provide custom body for ES search request and transformation of the result

### DISCUSSION THREAD

**Discussion by AurelPaulovic** [jertel/elastalert2#916] [Q&A] [answered]
I'm trying to use ElastAlert2 to alert on error logs. To do that I need to use a specific search request in ES that looks like this:
```json
{
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {"term": { "level": "ERROR" }},
        {
          "range": {
            "@timestamp": {
              "format": "strict_date_optional_time",
              "gte": "2022-07-27T17:40:00.000Z",
              "lte": "2022-07-27T18:00:00.000Z"
            }
          }
        }
      ]
    }
  },
  "aggs": {
    "application": {
      "terms": {
        "field": "application",
        "size": 100,
        "show_term_doc_count_error": true
      },
      "aggs": {
        "error": {
          "categorize_text": {
            "field": "message",
            "size": 100,
            "categorization_filters": [
              "[a-zA-Z0-9\\._\\-\\[\\]]{24,}"
            ]
          },
          "aggs": {
            "example": {
              "top_hits": {
                "size": 1,
                "_source": {
                  "includes": [ "application", "@timestamp", "message" ]
                }
              }
            }
          }
        }
      }
    }
  }
}
```
What the query does is that I filter on ERROR logs in some time window and then do 3 nested aggregations:
1.  the first one aggregates the error logs by `application`; 
2. the second one then tries to group all error messages from given application into separate buckets using similarity of the tokens so that it deduplicates logs that e.g. differ only by some ID, user name and such;
3. and finally for every error message bucket I select one example message.

I need to use these aggregations because applications might run rampant and write tens of thousands of duplicate error logs and trying to load them all into ElastAlert2 would be prohibitively expensive. So just running a simple search and then trying to use some custom Enhancement module that would do the fuzzy deduplication is not a viable solution.

ElastAlert2 currently does not support specifying a custom search request with arbitrary aggregations so I'm trying to patch it in a way, that would allow me to override the `ElastAlert.get_query` calls by providing a custom implementation in a custom Rule and subsequently preprocess the result returned from the search to massage it into a form expected by the implementation in `ElastAlert.get_hits`.

Is there any such feature planned? I could theoretically try to prepare a PR, but I'm not a python developer so I've got no idea what should a right python idiomatic implementation look like. Nor do I have the budget to make unit tests for it.

**Comment by jertel** [comment, marked as answer]
It is possible to use nested aggregations via the `query_key` list option. However, I'm unaware of support for using a categorized aggregation as in your example.

There are no plans to add such a feature. Since you do not have "budget" to supply the unit test coverage I suggest forking this project in order to make the changes you require.

Perhaps Elastic has an alternative solution, where you can build a "view" into an index, using your custom query, and then have ElastAlert2 query that "view" instead of an actual index. I know there is the "Index Alias" feature, which presents a psuedo index across one or more physical indices, with an optional filter to only return a subset of documents, but I don't know if something like that can support aggregations as well. This is a good question for the Elastic team, if you were interested in seeing if this feature exists.
